src/signing/collector/signatures_collector.rs

```diff
@@ -0,0 +1,554 @@
+                        .map(|f| f.factor_source_id())
+                        .collect(),
+                );
+                let response = interactor.sign(request).await?;
+                self.process_batch_response(response);
+            }
+
+            // Serial Interactor: One Factor Sources at a time
+            // After each factor source we pass the result to the collector
+            // updating its internal state so that we state about being able
+            // to skip the next factor source or not.
+            SigningInteractor::Serial(interactor) => {
+                for factor_source in factor_sources {
+                    // Prepare the request for the interactor
+                    let request =
+                        self.request_for_serial_interactor(&factor_source.factor_source_id());
+
+                    // Produce the results from the interactor
+                    let response = interactor.sign(request).await?;
+
+                    // Report the results back to the collector
+                    self.process_batch_response(response);
+
+                    if !self.continue_if_necessary()? {
+                        break;
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    async fn sign_with_factors_of_kind(
+        &self,
+        factor_sources_of_kind: FactorSourcesOfKind,
+    ) -> Result<()> {
+        let result = self.use_factor_sources(&factor_sources_of_kind).await;
+
+        match result {
+            Ok(_) => {}
+            Err(_) => self.process_batch_response(SignWithFactorSourceOrSourcesOutcome::Skipped {
+                ids_of_skipped_factors_sources: factor_sources_of_kind.factor_source_ids(),
+            }),
+        }
+        Ok(())
+    }
+
+    /// In decreasing "friction order"
+    async fn sign_with_factors(&self) -> Result<()> {
+        let factors_of_kind = self.dependencies.factors_of_kind.clone();
+        for factor_sources_of_kind in factors_of_kind.into_iter() {
+            if !self.continue_if_necessary()? {
+                break; // finished early, we have fulfilled signing requirements of all transactions
+            }
+            self.sign_with_factors_of_kind(factor_sources_of_kind)
+                .await?;
+        }
+        Ok(())
+    }
+
+    fn input_for_interactor(
+        &self,
+        factor_source_id: &FactorSourceIDFromHash,
+    ) -> BatchTXBatchKeySigningRequest {
+        self.state
+            .borrow()
+            .petitions
+            .borrow()
+            .input_for_interactor(factor_source_id)
+    }
+
+    fn request_for_serial_interactor(
+        &self,
+        factor_source_id: &FactorSourceIDFromHash,
+    ) -> SerialBatchSigningRequest {
+        let batch_signing_request = self.input_for_interactor(factor_source_id);
+
+        SerialBatchSigningRequest::new(
+            batch_signing_request,
+            self.invalid_transactions_if_skipped(factor_source_id)
```

Comment:
A request is created with the invalid transactions if skipped? hard to understand a bit, will see what this actual function does.

Reply:
The end user must know which transactinos would fail if she were to skip this factor source. So the impl of the interactor trait - the Host - will show this info to the user, "Are you sure you wanna skip signing with this factor, if you do these transactions would fail" (with good copy and UI).

Reply:
we would show a list of tx_ids that would fail? 
need to review the code further, but then in the case of skip, I imagine that we would want to strip out those transactions from all other further signature requests right?

Reply:
Right, good find! So the standard behaviour - controlled by the new: `SigningFinishEarlyStrategy` is to **finish early** if any transaction is invalid. I think that is what we want, but it is possible to specify `Continue` if some transaction is invalid. Then as you say, the current code would indeed display already failed tx. Hmm... gosh... I will give it a go at fixing this properly and will eval after some hours if it is worth it. I imagine it can be quite complex, since the proper solution is to refrain from even prompting using a factor source if it only involves already failed tx!

Reply:
> SigningFinishEarlyStrategy is to finish early if any transaction is invalid. I think that is what we want.

I am not sure this is what we want, user might not have the Ledger device to sign some transactions, but they would be happy to continue signing the rest which do not required the Ledger device, so the signing should not finish early, unless all of the transactions require the Ledger device.

Reply:
good that it is a single line of code to change :D

Reply:
fixed in https://github.com/radixdlt/sargon-mfa/pull/5